mllib/src/main/scala/org/apache/spark/ml/image/ImageSchema.scala

```diff
@@ -0,0 +1,236 @@
+      val hasAlpha = img.getColorModel.hasAlpha
+
+      val height = img.getHeight
+      val width = img.getWidth
+      val (nChannels, mode) = if (isGray) {
+        (1, ocvTypes("CV_8UC1"))
+      } else if (hasAlpha) {
+        (4, ocvTypes("CV_8UC4"))
+      } else {
+        (3, ocvTypes("CV_8UC3"))
+      }
+
+      val imageSize = height * width * nChannels
+      assert(imageSize < 1e9, "image is too large")
+      val decoded = Array.ofDim[Byte](imageSize)
+
+      // Grayscale images in Java require special handling to get the correct intensity
+      if (isGray) {
+        var offset = 0
+        val raster = img.getRaster
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            decoded(offset) = raster.getSample(w, h, 0).toByte
+            offset += 1
+          }
+        }
+      } else {
+        var offset = 0
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            val color = new Color(img.getRGB(w, h))
+
+            decoded(offset) = color.getBlue.toByte
+            decoded(offset + 1) = color.getGreen.toByte
+            decoded(offset + 2) = color.getRed.toByte
+            if (nChannels == 4) {
+              decoded(offset + 3) = color.getAlpha.toByte
+            }
+            offset += nChannels
+          }
+        }
+      }
+
+      // the internal "Row" is needed, because the image is a single DataFrame column
+      Some(Row(Row(origin, height, width, nChannels, mode, decoded)))
+    }
+  }
+
+  /**
+   * :: Experimental ::
+   * Read the directory of images from the local or remote source
+   *
+   * @param path Path to the image directory
+   * @param sparkSession Spark Session, if omitted gets or creates the session
+   * @param recursive Recursive path search flag
+   * @param numPartitions Number of the DataFrame partitions,
+   *                      if omitted uses defaultParallelism instead
+   * @param dropImageFailures Drop the files that are not valid images from the result
+   * @param sampleRatio Fraction of the files loaded
+   * @return DataFrame with a single column "image" of images;
+   *         see ImageSchema for the details
+   */
+  def readImages(
+      path: String,
+      sparkSession: SparkSession = null,
+      recursive: Boolean = false,
+      numPartitions: Int = 0,
+      dropImageFailures: Boolean = false,
+      sampleRatio: Double = 1.0): DataFrame = {
+    require(sampleRatio <= 1.0 && sampleRatio >= 0, "sampleRatio should be between 0 and 1")
+
+    val session = if (sparkSession != null) sparkSession else SparkSession.builder().getOrCreate
+    val partitions =
+      if (numPartitions > 0) {
+        numPartitions
+      } else {
+        session.sparkContext.defaultParallelism
+      }
+
+    RecursiveFlag.withRecursiveFlag(recursive, session) {
+      SamplePathFilter.withPathFilter(sampleRatio, session) {
+        val streams = session.sparkContext.binaryFiles(path, partitions).repartition(partitions)
```

Comment:
@jkbradley I think the `repartition` won't be expensive, because `.binaryFiles` returns `RDD[(String, PortableDataStream)]` which do not contain file content. (Only contain file split info in `PortableDataStream`.

Reply:
But, we should add comment that this is workround for the param ignored bug

Reply:
I think the point is, it triggers a shuffle and redistribute the data in some cases. This can potentially be expensive if there are many files.

Reply:
Yes, if there are large numbers of files it will still be expensive. But if `repartition` is postponed to be after image data loaded, then the cost will be much more expensive.
How about setting a default value of numPartitions = -1 , representing no `repartition` handling, other case keep current code.

Reply:
The last suggestion sounds good to me (default to no repartition, but do repartition if requested).

Reply:
good idea, I've set the default to -1 in both scala and python.  If the value is -1, we don't repartition in the scala code